pkg/runner/hash_map.go

```diff
@@ -0,0 +1,96 @@
+package runner
+
+import "sync"
+
+type hashMap struct {
+	mut    sync.RWMutex
+	lookup map[uint64][]Task
+}
+
+// newHashMap creates a new hashMap allocated to handle at least size unique
+// hashes.
+func newHashMap(size int) *hashMap {
+	return &hashMap{
+		lookup: make(map[uint64][]Task, size),
+	}
+}
+
+// Add adds the provided Task into the hashMap. It returns false if t
+// already exists in the hashMap.
+func (hm *hashMap) Add(t Task) bool {
+	hm.mut.Lock()
+	defer hm.mut.Unlock()
+
+	hash := t.Hash()
```

Comment:
Can this be simplified to use `Has` and move the locks to after the `Has` call?

Reply:
Technically no, if two `Add` calls run in parallel with the same task, there's a chance that the task gets added twice depending on if they both call `Has` at the same time before obtaining the lock.

Reply:
(But even if the lock was held for the entirety of Add, re-using Has here means the hash needs to get generated twice, so this performs better)